src/plot_api/plot_api.js

```diff
@@ -354,6 +361,10 @@ exports.plot = function(gd, data, layout, config) {
         initInteractions,
         Plots.addLinks,
         Plots.rehover,
+        // TODO: shouldn't really need doAutoMargin here, but without it some
+        // edits don't manage to tweak the margins the way they should
+        // in the tests, the only one that presently breaks is axis.position
+        Plots.doAutoMargin,
```

Comment:
This is a little unfortunate. This could be a big performance hit. Would it be too much to ask to try to remove this in this PR?

Reply:
If only axis.position is affected, we could add a special (hopefully temporary) relayout flag to call `Plots.doAutoMargin` after `finalDraw`

Reply:
I'll take a look again, didn't really understand why it was needed so I wasn't confident it was *only* `ax.position` - that's just the only one that we test directly that triggered it. Anyway it's a very minor performance hit if the margins *don't* change at that point.

Reply:
>  Anyway it's a very minor performance hit if the margins don't change at that point.

Oh right, `doAutoMargin` is smart<sup>TM</sup> like that. This additional `doAutoMargin` isn't a big then.

Reply:
On further investigation, this `doAutoMargin` call is definitely necessary for now. The reason is that axis automargin is not part of `marginPushers` - they reserve the *right* to push the margins there now with `Axes.allowAutoMargin` but the actual adjustment comes later during `Axes.doTicks`.

As discussed above, this `doAutoMargin` call won't actually redraw anything if the margins didn't change, so there's usually a minimal cost to this, but when the margins *do* change this triggers a full replot.

Long-term it would be far better to completely separate out margin calculations from drawing, so we *never* need to draw multiple times for a single change. But this addition doesn't represent a performance hit relative to the status quo.

Reply:
> The reason is that axis automargin is not part of marginPushers - they reserve the right to push the margins there now with Axes.allowAutoMargin but the actual adjustment comes later during Axes.doTicks.

Thanks for writing this down! Could update that comment in the code above this line to reflect these new findings?

> Long-term it would be far better to completely separate out margin calculations from drawing, so we never need to draw multiple times for a single change. But this addition doesn't represent a performance hit relative to the status quo.

The _dream_.

Reply:
Oh and about

> Long-term it would be far better to completely separate out margin calculations from drawing, so we never need to draw multiple times for a single change. But this addition doesn't represent a performance hit relative to the status quo.


Is there an issue about this already? If not, we should open one.

Reply:
comments improved in https://github.com/plotly/plotly.js/pull/2681/commits/230a782766ae9864eb5ac803c5fb069bedb9ab25, and new issue #2704 to track it.